### ISSUE TITLE
What is the best way to check user status right after every login ?

### ISSUE THREAD

**Issue by tech-a-go-go** [awinogrodzki/next-firebase-auth-edge#211]
Hi there

I want to check user's status like his/her stripe subscription is still active or something after every login or create my application side user account after first login.
If all status is fine, I want to redirect the user to the dashboard page.

I thought the following code is just fine at the first but not I guess...
It calls checkUserStatus() after fetching /api/login but a malicious user can skip checkUserStatus() by chrome dev tools since this code is using 'use client'.

```
'use client'
...
  async function handleLogin(credential: UserCredential) {
    const idToken = await credential.user.getIdToken();
    await fetch("/api/login", {
      headers: {
        Authorization: `Bearer ${idToken}`,
      },
    });
    
    checkUserStatus()  // checking user status here !! it's server action or api call
    
    router.push("/dashboard");
  }
```

Any suggestions how to be sure checking status will be done after login ?

Thanks a lot.

ps)
I don't want to check user status on all pages including dashboard. That will be messy.

**Comment by awinogrodzki** [comment]
Hey @tech-a-go-go!

I assume you want to redirect user to `/dashboard` just after they login, but only if they have correct credentials. What you are doing is correct.

> but a malicious user can skip checkUserStatus() by chrome dev tools since this code is using 'use client'.

You can solve this problem by introducing a guard inside `handleValidToken` in the middleware:

```tsx
    handleValidToken: async ({token, decodedToken, customToken}, headers) => {
      if (requrest.nextUrl.pathname === '/dashboard' && !hasUserAccess(decodedToken)) {
        return redirectToHome(request);
      } 

      // Authenticated user should not be able to access /login, /register and /reset-password routes
      if (PUBLIC_PATHS.includes(request.nextUrl.pathname)) {
        return redirectToHome(request);
      }

      return NextResponse.next({
        request: {
          headers
        }
      });
    },
```

**Comment by tech-a-go-go** [comment]
@awinogrodzki 

Thank you for your replay but since my English is poor, I couldn't explain what I really wanted to do.

Sorry but let me try again.

I need to create user data in my application's database upon the user's first login.

The login process is performed via a fetch request to /api/login:
```
await fetch("/api/login", {
      headers: {
        Authorization: `Bearer ${idToken}`,
      },
});
```

However, since /api/login is dynamically created by middleware, I am unable to insert my custom logic (creating user data in the app's DB) into this endpoint.

Could you please advise on how I can achieve user data creation on the first login?

Thank you for your assistance!

**Comment by awinogrodzki** [comment]
I understand your point @tech-a-go-go. At this moment, it's not possible to inject any additional logic to `/api/login` call. 
I will rethink this – maybe I could introduce some `handleLogin` option that will allow to decorate `/api/login` handler.

Until then, I think you can solve the problem by introducing custom login API handler:

This is basically how `/api/login` handler works:
```tsx
import {NextRequest, NextResponse} from 'next/server';
import {getFirebaseAuth} from 'next-firebase-auth-edge';
import {appendAuthCookies} from 'next-firebase-auth-edge/lib/next/cookies';

export async function performLogin(
  headers: Headers,
  options: SetAuthCookiesOptions
): Promise<NextResponse> {
  const {getCustomIdAndRefreshTokens} = getFirebaseAuth({
    serviceAccount: options.serviceAccount,
    apiKey: options.apiKey,
    tenantId: options.tenantId
  });
  const token = headers.get('Authorization')?.split(' ')[1] ?? '';

  if (!token) {
    const response = new NextResponse(
      JSON.stringify({success: false, message: 'Missing token'}),
      {
        status: 400,
        headers: {'content-type': 'application/json'}
      }
    );

    return response;
  }

  const appCheckToken = headers.get('X-Firebase-AppCheck') ?? undefined;
  const referer = getReferer(headers) ?? '';

  const customTokens = await getCustomIdAndRefreshTokens(token, {
    appCheckToken,
    referer
  });

  const response = new NextResponse(JSON.stringify({success: true}), {
    status: 200,
    headers: {'content-type': 'application/json'}
  });

  await appendAuthCookies(response, customTokens, options);

  return response;
}
```

In case you need some custom logic, I think you can create custom [API Route Handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) based on the code above

**Comment by tech-a-go-go** [comment]
@awinogrodzki 

Thank you for your response and the detailed explanation !

Until the handleLogin option is implemented, I will modify the performLogin API to handle the login requests via fetch. I will update the performLogin function to create user data if the authentication is successful and the user data does not already exist in the database.

Thank you for your assistance.
